src/NuGet.Core/NuGet.Commands/TrustedSignersCommand/TrustedSignerActionsProvider.cs

```diff
@@ -0,0 +1,310 @@
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException(Strings.ArgumentCannotBeNullOrEmpty, nameof(name));
+            }
+
+            if (!Enum.IsDefined(typeof(VerificationTarget), trustTarget) || trustTarget == VerificationTarget.Unknown)
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Error_UnsupportedTrustTarget, trustTarget.ToString()));
+            }
+
+            if (trustTarget == VerificationTarget.Author && owners != null && owners.Any())
+            {
+                throw new ArgumentException(Strings.Error_TrustedAuthorNoOwners);
+            }
+
+            var v3ServiceIndex = string.Empty;
+            IRepositorySignature reposig = null;
+            var trustingRepository = trustTarget.HasFlag(VerificationTarget.Repository);
+
+            var primarySignature = await package.GetPrimarySignatureAsync(token);
+
+            if (primarySignature == null)
+            {
+                throw new InvalidOperationException(Strings.Error_PackageNotSigned);
+            }
+
+            if (trustingRepository)
+            {
+                if (primarySignature.Type == SignatureType.Repository)
+                {
+                    reposig = primarySignature as RepositoryPrimarySignature;
+                }
+                else
+                {
+                    var countersignature = RepositoryCountersignature.GetRepositoryCountersignature(primarySignature);
+                    reposig = countersignature ?? throw new InvalidOperationException(Strings.Error_RepoTrustExpectedRepoSignature);
+                }
+
+                v3ServiceIndex = reposig.V3ServiceIndexUrl.AbsoluteUri;
+            }
+
+            var signers = _trustedSignersProvider.GetTrustedSigners();
+            foreach (var existingSigner in signers)
+            {
+                if (string.Equals(existingSigner.Name, name, StringComparison.Ordinal))
+                {
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Strings.Error_TrustedSignerAlreadyExists, name));
+                }
+
+                if (trustingRepository && existingSigner is RepositoryItem repoItem && string.Equals(repoItem.ServiceIndex, v3ServiceIndex, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Strings.Error_TrustedRepoAlreadyExists, v3ServiceIndex));
+                }
+            }
+
+            if (trustingRepository)
+            {
+                var certificateItem = GetCertificateItemForSignature(reposig, allowUntrustedRoot);
+
+                _trustedSignersProvider.AddOrUpdateTrustedSigner(new RepositoryItem(name, v3ServiceIndex, CreateOwnersList(owners), certificateItem));
+            }
+
+            if (trustTarget.HasFlag(VerificationTarget.Author))
```

Comment:
Could there be both `Author` and `Repository` flags in single trustTarget? if not, then may be if/else be more optimized

Reply:
Currently it's not supported. The reason I prefer having two if statements is to simplify the changes if in the future we want to support it. I don't have a strong opinion about this so I can change it.